sys/senml/phydat.c

```diff
@@ -0,0 +1,179 @@
+ * directory for more details.
+ */
+
+#include <stdint.h>
+
+#include "math.h"
+#include "senml.h"
+#include "senml/phydat.h"
+
+static uint8_t phydat_unit_to_senml_unit(uint8_t unit)
+{
+    switch (unit) {
+    /* Compatible units */
+    case UNIT_TEMP_C:  return SENML_UNIT_CELSIUS;
+    case UNIT_TEMP_K:  return SENML_UNIT_KELVIN;
+    case UNIT_LUX:     return SENML_UNIT_LUX;
+    case UNIT_M:       return SENML_UNIT_METER;
+    case UNIT_M2:      return SENML_UNIT_SQUARE_METER;
+    case UNIT_M3:      return SENML_UNIT_CUBIC_METER;
+    case UNIT_GR:      return SENML_UNIT_GRAM;
+    case UNIT_A:       return SENML_UNIT_AMPERE;
+    case UNIT_V:       return SENML_UNIT_VOLT;
+    case UNIT_W:       return SENML_UNIT_WATT;
+    case UNIT_COULOMB: return SENML_UNIT_COULOMB;
+    case UNIT_F:       return SENML_UNIT_FARAD;
+    case UNIT_OHM:     return SENML_UNIT_OHM;
+    case UNIT_PH:      return SENML_UNIT_PH;
+    case UNIT_PA:      return SENML_UNIT_PASCAL;
+    case UNIT_CD:      return SENML_UNIT_CANDELA;
+
+    /* Compatible Secondary units */
+    case UNIT_DBM:     return SENML_UNIT_DECIBEL_MILLIWATT;
+    case UNIT_PERCENT: return SENML_UNIT_PERCENT;
+    case UNIT_PERMILL: return SENML_UNIT_PERMILLE;
+    case UNIT_PPM:     return SENML_UNIT_PARTS_PER_MILLION;
+    case UNIT_PPB:     return SENML_UNIT_PARTS_PER_BILLION;
+
+    /* Incompatible units */
+    case UNIT_TEMP_F:  return SENML_UNIT_NONE;      /* use K or Cel instead */
+    case UNIT_GS:      return SENML_UNIT_NONE;      /* use T instead */
+    case UNIT_G:       return SENML_UNIT_NONE;      /* use m/s2 instead */
+    case UNIT_BAR:     return SENML_UNIT_NONE;      /* use Pa or hPa instead */
+    case UNIT_TIME:    return SENML_UNIT_NONE;      /* split into second/minute/hour */
+    case UNIT_DATE:    return SENML_UNIT_NONE;      /* split into day/month/year */
+    case UNIT_GPM3:    return SENML_UNIT_NONE;      /* use kg/m3 instead */
+    case UNIT_DPS:     return SENML_UNIT_NONE;      /* no alternative */
+    case UNIT_CPM3:    return SENML_UNIT_NONE;      /* no alternative */
+    default:           return SENML_UNIT_NONE;
+    }
+}
+
+void phydat_time_to_senml(senml_value_t *senml, const phydat_t *phydat)
+{
+    senml->value = senml_int((int32_t)phydat->val[0]
+                             + 60 * (int32_t)phydat->val[1]
+                             + 3600 * (int32_t)phydat->val[2]);
+    senml->attr.unit = SENML_UNIT_SECOND;
+}
+
+static int16_t month_to_yday[] = { 334, 0, 31, 59, 90, 120, 151,
+                                   181, 212, 243, 273, 304, 334 };
+
+void phydat_date_to_senml(senml_value_t *senml, const phydat_t *phydat)
+{
+    senml->attr.unit = SENML_UNIT_SECOND;
+
+    int64_t month = (int64_t)phydat->val[0];
+    int64_t year = (int64_t)phydat->val[0];
+
+    /* Calculate days of the year based on the month and leap year */
+    int64_t day = (int64_t)phydat->val[0] + month_to_yday[month] +
```

Comment:
something seems wrong here all information is in  phydat->val[0] ??

Reply:
There were some more errors in this part of the code. I have fixed it, and added a few test cases for `senml_phydat` including ones for unix time calculation.